pkg/providers/snow/snow.go

```diff
@@ -0,0 +1,242 @@
+package snow
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"time"
+
+	"github.com/aws/eks-anywhere/pkg/api/v1alpha1"
+	"github.com/aws/eks-anywhere/pkg/bootstrapper"
+	"github.com/aws/eks-anywhere/pkg/cluster"
+	"github.com/aws/eks-anywhere/pkg/constants"
+	"github.com/aws/eks-anywhere/pkg/filewriter"
+	"github.com/aws/eks-anywhere/pkg/providers"
+	"github.com/aws/eks-anywhere/pkg/retrier"
+	"github.com/aws/eks-anywhere/pkg/types"
+	releasev1alpha1 "github.com/aws/eks-anywhere/release/api/v1alpha1"
+)
+
+const (
+	eksaSnowCredentialsFileKey = "EKSA_SNOW_DEVICES_CREDENTIALS_FILE"
+	eksaSnowCABundlesFileKey   = "EKSA_SNOW_DEVICES_CA_BUNDLES_FILE"
+	snowCredentialsKey         = "AWS_B64ENCODED_CREDENTIALS"
+	snowCertsKey               = "AWS_B64ENCODED_CA_BUNDLES"
+	maxRetries                 = 30
+	backOffPeriod              = 5 * time.Second
+)
+
+var requiredEnvs = []string{
+	"AWS_ACCESS_KEY_ID", "AWS_SECRET_ACCESS_KEY", "AWS_REGION", "AWS_SESSION_TOKEN",
+}
+
+type snowProvider struct {
+	datacenterConfig      *v1alpha1.SnowDatacenterConfig
+	machineConfigs        map[string]*v1alpha1.SnowMachineConfig
+	clusterConfig         *v1alpha1.Cluster
+	providerKubectlClient ProviderKubectlClient
+	writer                filewriter.FileWriter
+	templateBuilder       *SnowTemplateBuilder
+	retrier               *retrier.Retrier
+	bootstrapCreds        bootstrapCreds
+}
+
+type ProviderKubectlClient interface{}
+
+func NewProvider(datacenterConfig *v1alpha1.SnowDatacenterConfig, machineConfigs map[string]*v1alpha1.SnowMachineConfig, clusterConfig *v1alpha1.Cluster, providerKubectlClient ProviderKubectlClient, writer filewriter.FileWriter, now types.NowFunc) *snowProvider {
+	return NewProviderCustomNet(
+		datacenterConfig,
+		machineConfigs,
+		clusterConfig,
+		providerKubectlClient,
+		writer,
+		now,
+	)
+}
+
+func NewProviderCustomNet(datacenterConfig *v1alpha1.SnowDatacenterConfig, machineConfigs map[string]*v1alpha1.SnowMachineConfig, clusterConfig *v1alpha1.Cluster, providerKubectlClient ProviderKubectlClient, writer filewriter.FileWriter, now types.NowFunc) *snowProvider {
+	var controlPlaneMachineSpec, etcdMachineSpec *v1alpha1.SnowMachineConfigSpec
+	if clusterConfig.Spec.ControlPlaneConfiguration.MachineGroupRef != nil && machineConfigs[clusterConfig.Spec.ControlPlaneConfiguration.MachineGroupRef.Name] != nil {
+		controlPlaneMachineSpec = &machineConfigs[clusterConfig.Spec.ControlPlaneConfiguration.MachineGroupRef.Name].Spec
+	}
+
+	workerNodeGroupMachineSpecs := make(map[string]v1alpha1.SnowMachineConfigSpec, len(machineConfigs))
+
+	if clusterConfig.Spec.ExternalEtcdConfiguration != nil {
+		if clusterConfig.Spec.ExternalEtcdConfiguration.MachineGroupRef != nil && machineConfigs[clusterConfig.Spec.ExternalEtcdConfiguration.MachineGroupRef.Name] != nil {
+			etcdMachineSpec = &machineConfigs[clusterConfig.Spec.ExternalEtcdConfiguration.MachineGroupRef.Name].Spec
+		}
+	}
```

Comment:
nit: I think we should find a way to not have to do all if this in the provider constructor. No need to change it but let's keep this convo going